File: Python/beginner/05_ShortestPath/boj_1753.py
import sys
import heapq

input = sys.stdin.readline

def solve(grp: list[list[tuple[int, int]]], start: int) -> list[int]:
    q = []
    distance = [int(1e9)] * (len(grp))
    
    distance[start] = 0
    heapq.heappush(q, (0, start))

    while q:
        dist, now = heapq.heappop(q)

        if distance[now] < dist:
            continue

        for end_vertex, weight in grp[now]:
            cost = distance[now] + weight

            if distance[end_vertex] > cost:
                distance[end_vertex] = cost
                heapq.heappush(q, (cost, end_vertex))

    return distance


if __name__ == '__main__':
    v, e = map(int, input().split())
    start = int(input())

    grp = [[] for _ in range(v + 1)]

    for _ in range(e):
        u, v, w = map(int, input().split())
        grp[u].append((v, w))

    distance = solve(grp, start)
    for i in range(1, len(distance)):
        if distance[i] == int(1e9):
            print("INF")
        else:
            print(distance[i])

File: Java/boj/Beginner/E_ShortestPath/boj_1753.java
package boj.Beginner.E_ShortestPath;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class boj_1753 {
    static StringTokenizer st;
    static List<Node>[] grp;

    static class Node implements Comparable<Node> {
        int end;
        int weight;

        public Node(int end, int weight) {
            this.end = end;
            this.weight = weight;
        }

        @Override
        public int compareTo(Node o) {
            return this.weight - o.weight;
        }
    }

    private static int[] dijkstra(int start) {
        PriorityQueue<Node> pq = new PriorityQueue<>();
        int[] distance = new int[grp.length];
        Arrays.fill(distance, Integer.MAX_VALUE);

        pq.add(new Node(start, 0));
        distance[start] = 0;

        while (!pq.isEmpty()) {
            Node tmp = pq.poll();
            int now = tmp.end;
            int dist = tmp.weight;

            if (distance[now] < dist)
                continue;

            for (Node node : grp[now]) {
                int end_vertex = node.end;
                int weight = node.weight;
                int cost = distance[now] + weight;

                if (distance[end_vertex] > cost) {
                    distance[end_vertex] = cost;
                    pq.add(new Node(end_vertex, cost));
                }
            }
        }

        return distance;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br =
                new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        st = new StringTokenizer(br.readLine());
        int V = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());
        int start = Integer.parseInt(br.readLine());

        grp = new ArrayList[V + 1];
        Arrays.setAll(grp, elem -> new ArrayList<>());

        while (E-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());

            grp[u].add(new Node(v, w));
        }

        int[] distance = dijkstra(start);
        for (int i = 1; i < distance.length; i++) {
            if (distance[i] == Integer.MAX_VALUE)
                sb.append("INF").append("\n");
            else
                sb.append(distance[i]).append("\n");
        }

        System.out.print(sb);
        br.close();
    }
}
